### DISCUSSION TITLE
v1.13.7

### DISCUSSION THREAD

**Discussion by deividAlfa** [deividAlfa/stm32_soldering_iron_controller#112] [Announcements]
- Longer beep on entering low power modes so the user can easily acknowledge without checking the screen.
- Workaround for crappy encoders, revert wake up/boost when a fast rotation->click sequence is detected, enter low power mode as intended.

<hr /><em>This discussion was created from the release <a href='https://github.com/deividAlfa/stm32_soldering_iron_controller/releases/tag/v1.13.7'>v1.13.7</a>.</em>

**Comment by jopl** [comment]
Dear Dave,
AMAZING !!! Thank you !!!
Just wanted to add czech language ... modified settings.h and gui_strings.c as well - this works well, czech language was added.
But ... we have got many diacritics in our language ... therefore I added missing glyphs in "...\Drivers\graphics\u8g2\tools\font" ("font_menu.bdf" and "font_small.bdf" - attached). Then used bdfconv.exe for conversion to c files (attached as well), hacked font sizes to 12/17 as per instruction and copied result strings to in "...\Drivers\graphics\u8g2\u8g2_fonts.c" (different folder then described in Notes.txt) ... suprisingly the "menu" font size decreased :-o. No need to modify u8g2.h in the same folder, as there is no size in array (???)
Then compiled without errors and ... unfortunately czech letters are missing :-( :-( :-(
Do I everythig well? And should I do something differently, please?
Once I compile the firmware successfully, I can share all the files with you (in case of your interest), of course.
Regards, Josef.
[Font_files.zip](https://github.com/deividAlfa/stm32_soldering_iron_controller/files/14394101/Font_files.zip)

**Reply by deividAlfa** [reply]
Did you extend the bat files?
https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/master/Drivers/graphics/u8g2/tools/font/bdfconv/make_font_menu.bat
https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/master/Drivers/graphics/u8g2/tools/font/bdfconv/make_font_small.bat

You need to add the new Unicode chars.

But sorry, I don't think I'll be adding more languages, it's a mess to maiNtain.
After a while, new options won't be translated anymore, original translators fly away.

**Reply by jopl** [reply]
Thanks so much for advise! You are absolutely right, adding the new Unicode chars codes into *.bat scripts was the crucial step.
OK, I added czech language just locally :-); it's up to you, what languages the FW will support.

**Reply by jopl** [reply]
Just for anyone, who is interested to have localized FW to czech language ...
Of course, follow link:
[Building the firmware](https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/master/Readme_files/Build.md)
And my final version of needed files:
[SolderingController_CZ_Fonts_Texts.zip](https://github.com/deividAlfa/stm32_soldering_iron_controller/files/14394399/SolderingController_CZ_Fonts_Texts.zip)

**Reply by deividAlfa** [reply]
Where're the modified BAT files? I added the translation but the bat must match the current requirements.

**Reply by jopl** [reply]
Dear Dave,
thanks so much! I really appreciate it. Attached is the most updated version (fonts, bats, strings, ...) ... The only limitation found is chinese font missing :-(. As I don't understand Chinese, I cannot recognize appropriate glyphs table :-(. Therefore I didn't update it in my fork ;-). Regards, Josef.
[CZ.zip](https://github.com/deividAlfa/stm32_soldering_iron_controller/files/14474500/CZ.zip)

**Reply by deividAlfa** [reply]
Thanks. That's sst311212 job, he made the fonts himself using some python magic.

**Comment by jopl** [comment]
Dear Dave,
I have got following tip for menu item "IRON" change:
I would change the word "IRON" for "PROFILE " (with space in the end) and modify code in
"static void SETTINGS_create(screen_t *scr)" routine in file "... Drivers\graphics\gui\screens\settings_screen.c" as  follows:

  // allocate memory for the concatenated string:
  char * StringProfile = (char*) malloc((strlen(profileStr[getCurrentProfile()])+ strlen(strings[lang].settings_IRON))*sizeof(char));
  strcpy(StringProfile, strings[lang].settings_IRON);
  strcat(StringProfile, profileStr[getCurrentProfile()]);
  newComboScreen(w, StringProfile, screen_iron, NULL);
  //newComboScreen(w, strings[lang].settings_IRON, screen_iron, NULL); // Original one

In this way, first menu item will look like for example "PROFILE T12" (or current profile name in the end), which seems more logical for me, as this settings are related to profile, isn't it?
BTW, except "Drag rotation", there is no way to change current profile, despite the "Operating Instructions" says:
"Switching between profiles can be done in the System menu" (in section IRON - first paragraph).
Regards, Josef.

**Reply by deividAlfa** [reply]
That was the old method - that description just happened to stay there, will be removed. 
I'm not sure "PROFILE T12" will fit in the screen. Maybe just "T12".

    newComboScreen(w, profileStr[getCurrentProfile()]), screen_iron, NULL);

Please understand supporting open source takes too much time and gives nothing back.
So, the activity of this project will be low, perhabs 1-2 hours a week at most.

Anyone is free to make a pull request after properly testing the code and ensuring it breaks nothing.

**Reply by jopl** [reply]
OK,
I am so sorry, just for evidence there is enough space to have full "PROFILE XXXX" - see pictures, despite the letter "E" is missing.
I'll test to pull request therefore.
Regards, Josef.

![PROFIL_C245](https://github.com/deividAlfa/stm32_soldering_iron_controller/assets/401876/590fb189-9bb0-4339-a596-21853f0d310f)
![PROFIL_T12](https://github.com/deividAlfa/stm32_soldering_iron_controller/assets/401876/0f8d3a34-b322-48cb-88ba-9119fed868c6)

**Reply by deividAlfa** [reply]
Don't be sorry, it's very welcome, but I'm currently a bit tired of this project. Will revisit this in the future!

**Comment by sst311212** [comment]
Sorry to bother you, but when editing default Cal250 value iron tip will heat up, in old firmware it won't do this.
For anyone that doing first calibration will not know which default value is fit for their iron tips.
It may burn out iron tip if default ADC is too high (especially for C210), and for tips safety, users should find the suit value for their tips in debug mode before editing default value.

https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/dc57501a7c8bf23fe3f0abd782de92593de434af/Drivers/graphics/gui/screens/calibration_screen.c#L101

**Reply by deividAlfa** [reply]
The debug mode is not for this, and it's never been intended to be required for firmware functionality.
This behavior is totally intended and explained in the manual, it's been like this for years except maybe for certain bugged releases, see [Calibration](https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/master/Readme_files/Operation.md#calibration).

There was a bug where it wouldn't wake the station up if it previously sleeping, see https://github.com/deividAlfa/stm32_soldering_iron_controller/commit/3d56cc72a2f2076d46d2ecccce444f31580b5b1c.

C210 gets about 1500 ADC for 400°C, current values are way lower.
The default values are pretty conservative to avoid  burning tips.

For safety... What the users should be doing is to read the manual before randomly changing things.

**Reply by sst311212** [reply]
Okay, I got it. Not heating up in previous firmware is a bug.😂

And for some clone stm will erase too much flash area, can't we add a variable to fixed erase page size? Or use flashTemp area to test proper page size at first boot? (write default temp to whole area and then erase 1 page, thus can find out real page size for this chip)

**Reply by deividAlfa** [reply]
I'm done with the freaking clones! 🤣
I already tried 1K and 2K flash sizes, they kept crashing, so I won't lose more time with them.

**Reply by sst311212** [reply]
Ok, because I don't have the clone one, otherwise I can figure out how's the flash memory will like after erasing one page.

Also very appreciate your hard working, this firmware works great so far. 👍

**Reply by deividAlfa** [reply]
Neither do I, people thinks I have half of AliExpress inventory and access to all Ksger catalog, while I just have my old 2019 Quicko!😂

**Reply by sst311212** [reply]
HAHA, you still trying make fake/clone stm32 works on custom firmware.
Really hope it can solve this problem with manually detection method. 🤣

**Reply by deividAlfa** [reply]
Yeah, I think those clones reporting "0xFFFF" flash size have different page sizes, some 1K, others 2K... It's the only possible explanation, as they were working well before the splitted flash partitions.

**Comment by godoto** [comment]
Maybe I'm doing something wrong, but where is the menu with t12/C245/C245 tip profiles? Has this been moved or is it no longer used? (v1.13.7 KSGER_v2_OLED)

**Reply by sst311212** [reply]
clockwise rotate with button pushed in main screen. (do it twice)

**Reply by deividAlfa** [reply]
I restored the profile selection from the menu: https://github.com/deividAlfa/stm32_soldering_iron_controller/commit/e434a15040e58b2941e949fb2c3eb36783b15072

**Reply by deividAlfa** [reply]
Chinese translation got nuked after adding Czech. I need you to properly add the Chinese Unicode characters here:

https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/master/Drivers/graphics/u8g2/tools/font/bdfconv/make_font_menu.bat
https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/master/Drivers/graphics/u8g2/tools/font/bdfconv/make_font_small.bat

And:
https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/master/Drivers/graphics/u8g2/tools/font/bdf/font_menu.bdf
https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/master/Drivers/graphics/u8g2/tools/font/bdf/font_small.bdf

Or the tool to do it (Where's did the python script go?)

I need it done properly, not depending on 3rd parties to update the fonts everytime something is changed.

How does the Chinese translation use a whooping 10KB of flash?
Using 8% of the entire flash is a lot and we're getting tight on space, only 10KB flash left with Chinese added.
It seems a lot to me, are you adding only the required glyphs? Or the entire Simplified Chinese charset?

**Reply by sst311212** [reply]
Yes, I only add needed glyphs into u8g2_fonts.c.
I desire not to touch non-chinese source, so I add some script to help generating chinese bdf and then combine them.
My tool is located at ```Drivers\graphics\u8g2\tools\font\tchinese```, use make_u8g2.bat to generate u8g2_aio.c.

**Reply by deividAlfa** [reply]
**So any new characters must be added here?**
https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/7a5fe2359b280b4a82214f2950ea7267c189b911/Drivers/graphics/u8g2/tools/font/tchinese/make_u8g2.bat#L8-L9

<br>

**Instead here?** (So these files are redundant and must be deleted?)
https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/7a5fe2359b280b4a82214f2950ea7267c189b911/Drivers/graphics/u8g2/tools/font/bdfconv/make_font_menu.bat#L3
https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/7a5fe2359b280b4a82214f2950ea7267c189b911/Drivers/graphics/u8g2/tools/font/bdfconv/make_font_small.bat#L3

**Reply by sst311212** [reply]
Yes, but I prefer keeping original files there and only combine map to my script.
It should help anybody add and test another glyphs more easier.

**Reply by jopl** [reply]
Hi,
if I am right, the result is attached (bat as well as u8g2_aio.c). Thanks for advises sst311212!!!
Regards, Josef.
[CZ_Tchinese.zip](https://github.com/deividAlfa/stm32_soldering_iron_controller/files/14485064/CZ_Tchinese.zip)

**Reply by jopl** [reply]
I have just found ... the menu has 3 lines only instead of 4 previously visible :-o
This process is reversible by compilation with u8g2_fonts.c without chinese characters, hmmm ...
J.

**Reply by sst311212** [reply]
You didn't change 10th byte, please read Notes.txt in bdfconv folder.

**Reply by jopl** [reply]
Dear Dave, you are right, this is THE REASON. Thx. Josef.
Corrected attached.
[CZ_Chinese_u8g2_fonts.zip](https://github.com/deividAlfa/stm32_soldering_iron_controller/files/14486860/CZ_Chinese_u8g2_fonts.zip)

**Comment by deividAlfa** [comment]
Having two font building scripts make no sense at all, so I unified them.

@sst311212 Please check if everything is correct:
- Moved the command for making the big iron temperature font to `make_u8g2.bat` and modified the `copy /b` line.
 https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/d89bc4d8c079763d5927516b292803ed7e4e89d4/Drivers/graphics/u8g2/tools/font/generate_fonts/make_u8g2.bat#L9-L12

- Wrote a proper [README](https://github.com/deividAlfa/stm32_soldering_iron_controller/blob/d89bc4d8c079763d5927516b292803ed7e4e89d4/Drivers/graphics/u8g2/tools/font/generate_fonts/README.md).

**Reply by sst311212** [reply]
Seems no errors here.
